### PR TITLE
Fix DruidNavigator loading multiple tables

### PR DESCRIPTION
**Pull request by kevinhinterlong**
It appears the locking mechanism was forced to time out while loading
multiple tables because the lock for getting all the tables was still active.

Now only the lock for getting all the tables (`loadAllDatasources()`) is used.

### CODE REVIEW

**Review by cdeszaq — DISMISSED** [review]
A few things here:

1. Changelog entry needed (probably in the `Fixes` section)
2. It would be nice to have a test in place that shows the issue and that this change fixes it.
3. I'm not sure I understand how this is fixing the problem. The main change I see is to move the `get` to the caller. I don't really understand how that solves the problem which, as I understand it, is that the follow-on calls are trying to run while the 1st call isn't finished. Given how the AsyncHttpClient works, multiple calls in flight at once shouldn't be a problem. And there's no lock involved (other than the implicit one in `Future::get`), so I'm a little confused there as well.
4. Don't we still want to have the later calls timeout or at least finish executing before the rest of the application starts up? I think this change would allow the app to start w/o having any of the tables configured...

File: fili-generic-example/src/main/java/com/yahoo/wiki/webservice/data/config/dimension/GenericDimensionConfigs.java [hunk 1]
```diff
@@ -17,39 +17,46 @@
...
+import java.util.HashMap;
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
 
 /**
  * Hold all the dimension configurations for a generic druid configuration.
  */
 public class GenericDimensionConfigs {
-    private final Set<DimensionConfig> dimensionConfigs;
+    private final Map<String, Set<DimensionConfig>> dimensionConfigs;
```
Comment (michael-mclawhorn) [inline fili-generic-example/src/main/java/com/yahoo/wiki/webservice/data/config/dimension/GenericDimensionConfigs.java]: It's not clear what the String in this map is.

File: fili-generic-example/src/test/groovy/com/yahoo/wiki/webservice/web/endpoints/MultipleDatasourceDruidConfigLoaderSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        druidWebService = new TestDruidWebService("testInstance")
+        druidWebService.jsonResponse = {
+            if (druidWebService.lastUrl.equals("/datasources/")) {
+                return new ObjectMapper().writeValueAsString(datasources)
+            } else if (druidWebService.lastUrl.equals("/datasources/table1/?full")) {
+                return getFullTable(datasources[0], table1_metrics, table1_dimensions)
+            } else if (druidWebService.lastUrl.equals("/datasources/table2/?full")) {
+                return getFullTable(datasources[1], table2_metrics, table2_dimensions)
+            }
+            println "No response for " + druidWebService.lastUrl
```
Comment (QubitPi) [inline fili-generic-example/src/test/groovy/com/yahoo/wiki/webservice/web/endpoints/MultipleDatasourceDruidConfigLoaderSpec.groovy]: Maybe you forgot to delete it?

File: fili-generic-example/src/test/groovy/com/yahoo/wiki/webservice/web/endpoints/AutomaticDruidConfigLoaderSpec.groovy [hunk 3]
```diff
@@ -53,7 +52,7 @@ public class AutomaticDruidConfigLoaderSpec extends Specification {
             } else if (druidWebService.lastUrl == '/datasources/' + datasource + "/?full") {
                 return expectedMetricsAndDimensions
             }
-            return "Unexpected URL"
+            println "No response for " + druidWebService.lastUrl
```
Comment (QubitPi) [inline fili-generic-example/src/test/groovy/com/yahoo/wiki/webservice/web/endpoints/AutomaticDruidConfigLoaderSpec.groovy]: Maybe you forgot to delete this?

File: fili-generic-example/src/main/java/com/yahoo/wiki/webservice/data/config/dimension/GenericDimensionConfigs.java [hunk 4]
```diff
@@ -17,39 +17,46 @@
...
+                                    )
+                            ).collect(
+                                    Collectors.collectingAndThen(
+                                            Collectors.toSet(),
+                                            Collections::unmodifiableSet
+                                    ));
+
+                    dimensionConfigs.put(dataSourceConfiguration.getName(), tableDimensionConfigs);
+                });
+        ;
```
Comment (QubitPi) [inline fili-generic-example/src/main/java/com/yahoo/wiki/webservice/data/config/dimension/GenericDimensionConfigs.java]: What's this for?

File: fili-generic-example/src/main/java/com/yahoo/wiki/webservice/data/config/dimension/GenericDimensionConfigs.java [hunk 5]
```diff
@@ -58,7 +65,21 @@ public GenericDimensionConfigs(Supplier<List<? extends DataSourceConfiguration>>
...
+        return dimensionConfigs.values().stream()
+                .flatMap(Set::stream)
+                .distinct()
+                .collect(Collectors.toSet());
+    }
+
+    /**
+     * Get all the dimension configurations associated with this datasource.
+     *
+     * @param dataSourceConfiguration the datasource configuration's dimensions to load
```
Comment (QubitPi) [inline fili-generic-example/src/main/java/com/yahoo/wiki/webservice/data/config/dimension/GenericDimensionConfigs.java]: Two spaces before description. Capitalize `the`

File: fili-generic-example/src/main/java/com/yahoo/wiki/webservice/data/config/table/GenericTableLoader.java [hunk 6]
```diff
@@ -21,29 +21,23 @@
...
-
 import org.joda.time.DateTimeZone;
 
-import java.util.Collections;
-import java.util.HashSet;
-import java.util.LinkedHashSet;
-import java.util.List;
-import java.util.Set;
+import javax.validation.constraints.NotNull;
+import java.util.*;
```
Comment (QubitPi) [inline fili-generic-example/src/main/java/com/yahoo/wiki/webservice/data/config/table/GenericTableLoader.java]: Avoid start import

File: fili-generic-example/src/test/groovy/com/yahoo/wiki/webservice/web/endpoints/MultipleDatasourceDruidConfigLoaderSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+                    "type": "none"
+                },
+            "binaryVersion": 9,
+            "size": 5537610,
+            "identifier": "${name}_2015-09-12T00:00:00.000Z_2015-09-13T00:00:00.000Z_2017-02-27T03:06:09.422Z"
+        }
+    ]
+}
+
+"""
```
Comment (QubitPi) [inline fili-generic-example/src/test/groovy/com/yahoo/wiki/webservice/web/endpoints/MultipleDatasourceDruidConfigLoaderSpec.groovy]: Align indentation

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by michael-mclawhorn — APPROVED** [review]
Looks pretty good

**Review by ghost — APPROVED** [review]

**Review by cdeszaq — COMMENTED** [review]
Partial review. Didn't get as far as `GenericDimensionConfigs`.

File: fili-generic-example/README.md [hunk 8]
```diff
@@ -90,4 +95,6 @@ Here are some sample queries that you can run to verify your server:
 
 1. In IntelliJ, go to `File -> Open`
 2. Select the `pom.xml` file at the root of the project
-3. Run `GenericMain` which can be found in `fili-generic-example` (e.g. right click and choose run)
+3. Under `src/main/resources/applicationConfig.properties`, change `bard__non_ui_druid_broker`, 
+`bard__ui_druid_broker`, `bard__druid_coord`, and other properties as needed.
```
Comment (kevinhinterlong) [inline fili-generic-example/README.md]: Yep these are set correctly for running it locally and they work with the wikipedia walkthrough. 

I added the following note in `README.md` to clarify:

> **NOTE:** if you're running this locally and haven't changed any settings (like the Wikipedia example) you can **skip step 3**.

**Comment by kevinhinterlong** [comment]
I added a note in the readme of the version this was last tested with. 
I think this should be good to merge now.

**Review by garyluoex — APPROVED** [review]
Rebase, and Im ok with it

**Pull request merged**